arch/risc-v/src/mpfs/mpfs_fpga_canfd.c

```diff
@@ -0,0 +1,3181 @@
+          putreg32(icr, priv->base + MPFS_CANFD_INT_MASK_SET_OFFSET);
+          putreg32(icr, priv->base + MPFS_CANFD_INT_STAT_OFFSET);
+
+          /* Main receiving routine */
+
+          mpfs_receive(priv);
+        }
+
+      /* TXT Buffer HW Command Interrupt */
+
+      if (isr & MPFS_CANFD_INT_STAT_TXBHCI)
+        {
+          /* Clear TX interrupt flags */
+
+          mpfs_txdone(priv);
+
+          /* Schedule work to poll for next available tx frame from the
+           * network.
+           */
+
+          work_queue(CANWORK, &priv->irqwork, mpfs_txdone_work, priv, 0);
+        }
+
+      /* Error Interrupts */
+
+      if (isr & MPFS_CANFD_INT_STAT_EWLI ||
+          isr & MPFS_CANFD_INT_STAT_FCSI ||
+          isr & MPFS_CANFD_INT_STAT_ALI ||
+          isr & MPFS_CANFD_INT_STAT_BEI)
+        {
+          icr = isr & (MPFS_CANFD_INT_STAT_EWLI |
+                      MPFS_CANFD_INT_STAT_FCSI |
+                      MPFS_CANFD_INT_STAT_ALI |
+                      MPFS_CANFD_INT_STAT_BEI);
+          caninfo("Some error interrupts. Clearing 0x%08x\n", icr);
+          putreg32(icr, priv->base + MPFS_CANFD_INT_STAT_OFFSET);
+          mpfs_err_interrupt(priv, isr);
+        }
+    }
+
+  /* Now, it seems that there are still some interrupt flags that remain
+   * stuck in INT_STAT reg.
+   */
+
+  canerr("Stuck interrupt (isr=%08x)\n", isr);
+
+  /* Check if any of the stuck one belongs to txb status. */
+
+  if (isr & MPFS_CANFD_INT_STAT_TXBHCI)
+    {
+      caninfo("txb_head=0x%08x txb_tail=0x%08x\n", priv->txb_head,
+              priv->txb_tail);
+      for (int i = 0; i < priv->ntxbufs; i++)
+        {
+          uint32_t status = mpfs_can_get_tx_status(priv, i);
+          caninfo("txb[%d] txb status=0x%08x\n", i, status);
+        }
+
+      /* Clear txb status change interrupt */
+
+      putreg32(MPFS_CANFD_INT_STAT_TXBHCI,
+               priv->base + MPFS_CANFD_INT_STAT_OFFSET);
+    }
+
+  /* Check if any of the stuck one belongs to RX buffer data overrun */
+
+  if (isr & MPFS_CANFD_INT_STAT_DOI)
+    {
+      struct canfd_frame *cf = (struct canfd_frame *)priv->rxdesc;
+
+      caninfo("rx fifo overflow\n");
+
+      cf->can_id = CAN_ERR_CRTL;
+      cf->data[1] = CAN_ERR_CRTL_RX_OVERFLOW;
+
+      /* Copy the buffer pointer to priv->dev..  Set amount of data
+       * in priv->dev.d_len
+       */
+
+      priv->dev.d_len = sizeof(struct canfd_frame);
+      priv->dev.d_buf = (uint8_t *)cf;
+
+      /* Send to socket interface */
+
+      NETDEV_RXPACKETS(&priv->dev);
+      can_input(&priv->dev);
+
+      /* Point the packet buffer back to the next Tx buffer that will be
+       * used during the next write.
+       */
+
+      priv->dev.d_buf = (uint8_t *)priv->txdesc;
+
+      /* Clear Data Overrun interrupt */
+
+      putreg32(MPFS_CANFD_INT_STAT_DOI,
+               priv->base + MPFS_CANFD_INT_STAT_OFFSET);
+    }
+
+  /* Clear and reset all interrupt. */
+
+  caninfo("Reset all interrupts...\n");
```

Comment:
I don't quite understand why the interrupt routine has to manually disable the interrupt source, will this not result in missed interrupts ? Where does the source get re-enabled ?